kernel/kernel-api/src/test/scala/io/delta/kernel/SnapshotManagerSuite.scala

```diff
@@ -93,4 +103,689 @@ class SnapshotManagerSuite extends AnyFunSuite {
+    )
+    testExpectedError[IllegalStateException](
+      fileList,
+      startCheckpoint = Optional.of(10),
+      versionToLoad = Optional.of(10),
+      expectedErrorMessageContains = "Could not find any delta files for version 10"
+    )
+  }
+
+  test("getLogSegmentForVersion: corrupted log missing json files / no way to construct history") {
+    def expectedErrorMessage(v: Int): String = {
+      s"""|Log file not found.
+          |Expected: ${FileNames.deltaFile(logPath, 0)}
+          |Found: ${FileNames.deltaFile(logPath, v)}""".stripMargin
+    }
+    testExpectedError[RuntimeException](
+      deltaFileStatuses(1L until 10L),
+      expectedErrorMessageContains = expectedErrorMessage(1)
+    )
+    testExpectedError[RuntimeException](
+      deltaFileStatuses(15L until 25L) ++ singularCheckpointFileStatuses(Seq(20)),
+      versionToLoad = Optional.of(17),
+      expectedErrorMessageContains = expectedErrorMessage(15)
+    )
+    testExpectedError[RuntimeException](
+      deltaFileStatuses(15L until 25L) ++ singularCheckpointFileStatuses(Seq(20)),
+      startCheckpoint = Optional.of(20),
+      versionToLoad = Optional.of(17),
+      expectedErrorMessageContains = expectedErrorMessage(15)
+    )
+    testExpectedError[RuntimeException](
+      deltaFileStatuses((0L until 5L) ++ (6L until 9L)),
+      expectedErrorMessageContains = "are not continuous"
+    )
+    // corrupt incomplete multi-part checkpoint
+    val corruptedCheckpointStatuses = FileNames.checkpointFileWithParts(logPath, 10, 5).asScala
+      .map(p => FileStatus.of(p.toString, 10, 10))
+      .take(4)
+    val deltas = deltaFileStatuses(10L to 13L)
+    testExpectedError[RuntimeException](
+      corruptedCheckpointStatuses ++ deltas,
+      Optional.empty(),
+      Optional.empty(),
+      expectedErrorMessageContains = expectedErrorMessage(10)
+    )
+  }
+
+  test("getLogSegmentForVersion: corrupt log but reading outside corrupted range") {
+    testNoCheckpoint(
+      (0L until 5L) ++ (6L until 9L),
+      Optional.of(4)
+    )
+    testWithSingularAndMultipartCheckpoint(
+      15L until 25L,
+      Seq(20),
+      versionToLoad = Optional.of(22)
+    )
+    testWithSingularAndMultipartCheckpoint(
+      15L until 25L,
+      Seq(20),
+      startCheckpoint = Optional.of(20),
+      versionToLoad = Optional.of(22)
+    )
+  }
+
+  test("getLogSegmentForVersion: corrupt _last_checkpoint (is after existing versions)") {
+    // in the case of a corrupted _last_checkpoint we revert to listing from version 0
+    // (on first run newFiles.isEmpty() but since startingCheckpointOpt.isPresent() re-list from 0)
+    testWithSingularAndMultipartCheckpoint(
+      (0L until 25L),
+      Seq(10L, 20L),
+      startCheckpoint = Optional.of(30)
+    )
+  }
+
+  // TODO recover from missing checkpoint (getLogSegmentWithMaxExclusiveCheckpointVersion)
+  test("getLogSegmentForVersion: corrupt _last_checkpoint refers to in range version " +
```

Comment:
Does Delta-Spark handle this case where the checkpoint file referred by the last checkpoint file doesn't exist or part of it missing?

Reply:
yes, `getLogSegmentWithMaxExclusiveCheckpointVersion` in SnapshotManager just needs to be implemented (should be mostly a copy from the spark counter-part)